eclipse-theia/theia
# GH-3985: Support for css/html via VS Code extension

#### What it does


Removed the CSS/HTML integration from monaco, and uses the corresponding VS Code extensions instead.

#### How to test
VS Code built-in HTML support:
 - [ ] diagnostics,
   - Check that you can produce and invalid HTML by defining a completely invalid `script`. Try `<script>totoal...invalid = $Hello JavaScript!$;</script>`, you should see the error in the Problems view too. You can also check the diagnostics for invalid HTML: `<html><script></scaript></html>`.
![Screen Shot 2020-06-08 at 13 44 08](https://user-images.githubusercontent.com/1405703/84026744-289c3f00-a98e-11ea-9e8a-353d3a4256fc.jpg)
 - [ ] snippet support, proposes completion for the `'.'`, `':'`, `'<'`, `'“'`, `'='`, and `'/'` chars,
   - User-level snippets are not yet supported: https://github.com/eclipse-theia/theia/issues/3686.
 - [ ] completion,
   - Open an HTML file type `<` see the tag completions.
![Screen Shot 2020-06-08 at 13 34 03](https://user-images.githubusercontent.com/1405703/84026023-d7d81680-a98c-11ea-9ac4-aaabb66abb19.jpg)
 - [ ] hover,
   - Open `/Users/akos.kitta/git/theia/packages/plugin-ext/src/main/browser/webview/pre/index.html` hover over `name` in `meta` tag, see the hover.
![Screen Shot 2020-06-08 at 13 33 18](https://user-images.githubusercontent.com/1405703/84025915-a5c6b480-a98c-11ea-9400-b1bd43d698c5.jpg)
 - [ ] document symbols,
    - Verify you see document symbols for the current HTML document (`Ctrl/Cmd+Shift+O`)
![Screen Shot 2020-06-08 at 13 57 41](https://user-images.githubusercontent.com/1405703/84027994-34890080-a990-11ea-80ce-256e78ed978a.jpg)
 - [ ] document highlight,
   - In a HTML file write the followings: `<div><div></div></div>`, select the inner `div`, verify that the only corresponding closing `div` is highlighted, not all of them.
![Screen Shot 2020-06-08 at 13 37 52](https://user-images.githubusercontent.com/1405703/84026254-45844280-a98d-11ea-9c3f-9da1df97b31a.jpg)
 - [ ] definitions,
   - You can use a `script` tag if you want to verify the definitions, check both `Go to Definition` and `Peek > Peek Definition`. Both should work. Here is an example: `<script>const x = 'foo'; console.log(x);</script>`.
![Screen Shot 2020-06-08 at 14 16 07](https://user-images.githubusercontent.com/1405703/84029409-9fd3d200-a992-11ea-9d1a-a34377fba7c0.jpg)
 - [ ] signature helper (triggered by `'('`),
   - Verify the signature helper, create a `script` tag, add a function, invoke it
![Screen Shot 2020-06-08 at 14 07 27](https://user-images.githubusercontent.com/1405703/84028705-80887500-a991-11ea-814e-4aa15f2da64a.jpg)
 - [ ] references,
   - Create a `script` tag and verify the JS references. Similar to `defintions`.
 - [ ] range folding,
   - Verify that you can expand/collapse ranges.
![screencast 2020-06-08 13-54-37](https://user-images.githubusercontent.com/1405703/84027768-d2300000-a98f-11ea-8720-64ca2dccc991.gif)
 - [ ] range formatting.
   - Make sure you can format a subset of the document. Open an HTML, select an unformatted range, format is. Formatting the selection will require both `'textDocument/foldingRange'` and `'textDocument/rangeFormatting'`. You can enable tracing for HTML via the settings.
![screencast 2020-06-08 13-48-10](https://user-images.githubusercontent.com/1405703/84027432-40c08e00-a98f-11ea-8ff4-21ec83f887fa.gif)
 - [ ] selection range,
   - See _range formatting_ 👆 
 - [ ] rename, and
   - Verify rename. Rename a tag to something.
![screencast 2020-06-08 14-09-14](https://user-images.githubusercontent.com/1405703/84028872-c5141080-a991-11ea-9599-efac2d9d636c.gif)
 - [ ] semantic highlighting.
   -  No idea how to verify this.

VS Code built-in CSS/SCSS/Less support:
 - [ ] diagnostics,
   - Check that you can generate errors in an invalid CSS file, the corresponding error shows up in the Problems view. Try `{{}}` in the editor and see the error: `at-rule or selector expected file:///Users/akos.kitta/git/theia/a.css`. Check for warnings, write `p > div { }` without rulesets, see the `Do not use empty rulesets` warning in the Problems view. You can disable validation by adding `"css.validate": false` to your `settings.json`.
 - [ ] completion,
   - You can see the default `#region` and `#endregion` completions in a CSS file. Also, type `p > div { d<|> }` (where `<|>` is the cursor position) you should see `display: grid|inline;` completion.
 - [ ] snippet support, proposes completion for the `'/'` and `'-'` chars,
   - User-level snippets are not yet supported: https://github.com/eclipse-theia/theia/issues/3686.
 - [ ] hover,
   - Check that hover works, the easiest is by hovering over an `url`  and opening the file in an editor. For example: `--theia-debug-start: url('start-inverse.svg')`.
 - [ ] document symbol,
   - Open a non-empty CSS file, press `Ctrl/Cmd+Shift+O` you can see the selectors defined in the CSS file.
 - [ ] references,
   - Right-click on a selector, you can trigger `Find All References`, the `Refrences` view opens and shows your selector.
 - [ ] definitions,
   - It's only partially [supported](https://code.visualstudio.com/Docs/languages/CSS#_go-to-declaration-and-find-references) for CSS. Open `packages/core/src/browser/style/menus.css`, right-click on the `--theia-private-menubar-height` usesage in the `.p-MenuBar-item` selector, pick `Go to Definition` from the context menu.
![screencast 2020-06-08 12-11-30](https://user-images.githubusercontent.com/1405703/84019270-5464f800-a981-11ea-9b2a-09aedc68c86e.gif)
 - [ ] document highlight,
   - Open `tree.css` in Theia, set the cursor after the first occurrence of `.theia-Tree` in the CSS file, see that other occurrences are highlighted in the file.
 - [ ] code actions,
   - Although code actions are not defined for CSS, you should see the enabled `Source Actions...` context menu item (with the browser example) from a CSS editor. Executing it will result in a `No source actions available` in the editor.
 - [ ] rename,
   - You can rename a selector in a CSS file. Open `tree.css`, for example, and rename `.theia-Tree` with the context menu (or `F2`). You can rename the symbol in the file. It won't update other occurrences outside of the current CSS file.
 - [ ] range folding,
   -  Sorround one of the CSS selectors with `/* #region */` and `/* #endregion */`. Add `"[css]": { "editor.foldingStrategy": "indentation" }` into the `settings.json` file, see in the editor, you can expand and fold ranges
![screencast 2020-06-08 11-59-25](https://user-images.githubusercontent.com/1405703/84017899-a3119280-a97f-11ea-95c4-d6d9cddeb89c.gif)
 - [ ] document link, and
   - You can open a `url('someLink.css')` in another editor.
 - [ ] selection range.
   - I have no idea how to test this.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
